compiler/cli/cli-common/src/org/jetbrains/kotlin/cli/common/repl/ReplApi.kt

```diff
@@ -111,7 +114,9 @@ interface ReplEvalAction {
 
 sealed class ReplEvalResult : Serializable {
     class ValueResult(val value: Any?) : ReplEvalResult() {
-        override fun toString(): String = "Result: $value"
+        val type = LastInferredTypeHolder.inferredType.get()
```

Comment:
I think you can obtain this type from `compileResult` at `GenericReplEvaluator.eval` and pass to `ValueResult`'s constructor here